<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <title>async函数</title>
    </head>
    <body>
        <!--
        async函数(源自ES2017)
        概念： 真正意义上去解决异步回调的问题，同步流程表达异步操作
        本质： Generator的语法糖
        语法：
            async function foo(){
                await 异步操作;
                await 异步操作；
            }
        特点：
            1、不需要像Generator去调用next方法，遇到await等待，当前的异步操作完成就往下执行
            2、返回的总是Promise对象，可以用then方法进行下一步操作
            3、async取代Generator函数的星号*，await取代Generator的yield
            4、语意上更为明确，使用简单，经临床验证，暂时没有任何副作用
        -->
        <script type="text/javascript" src="./js/jquery-1.10.1.min.js"></script>
        <script type="text/javascript">
            // async 基本使用
            async function foo() {
                // 只用 resolve，不用 reject，可以省略一个小括号
                return new Promise(resolve => {
                    /*
                    setTimeout(function() {
                        resolve();
                    }, 2000)
                    */

                    // 和上面的一模一样，更简洁
                    setTimeout(resolve, 2000);
                });
            }

            async function test() {
                console.log("开始执行", new Date().toTimeString());
                await foo(); // await 等待foo() 执行完毕
                console.log("执行完毕", new Date().toTimeString());
            }

            // async 函数不用 next，next，一调用就直接执行了
            test();

            // async 里面 await 的返回值
            function test2() {
                return "xxx";
            }
            async function asyncPrint() {
                let result = await test2();
                console.log(result);
            }
            // 如果是普通函数，await 得到的返回值就是函数执行的返回值
            asyncPrint();

            async function asyncPrint2() {
                // Promise 简写方法，生成一个promise对象，并且是成功的状态
                // let result = await Promise.resolve();  // 默认是 undefined
                let result = await Promise.resolve("aaa"); // 这就是 aaa
                console.log(result);
                result = await Promise.reject("失败啊！"); // 这里会异常，异常信息是 失败啊！
                console.log(result);
            }
            asyncPrint2();

            // 获取新闻内容
            async function getNews(url) {
                return new Promise((resolve, reject) => {
                    $.ajax({
                        method: "GET",
                        // 同名属性省略不写
                        // 等同于：url: url
                        url,
                        /*
                        success: function(data) {
                            resolve();
                        },
                        */
                        // 简化版
                        success: data => resolve(data),
                        // error: error => reject()
                        error: function(error) {
                            // 这里不用 reject()，而使用 resolve()，保证下面代码顺利执行
                            // reject();
                            resolve(false);
                        }
                    });
                });
            }

            async function sendXml() {
                let result = await getNews("http://localhost:3000/news?id=7");
                console.log(result);
                if (!result) {
                    alert("暂时没有新闻推送");
                    return;
                }
                result = await getNews("http://localhost:3000" + result.commentsUrl);
                console.log(result);
            }
            sendXml();
        </script>
    </body>
</html>
